csharp/SecureMemory/PlatformNative/LP64/Linux/OpenSSLCrypto.cs

```diff
@@ -0,0 +1,176 @@
+        {
+            byte[] outLengthBuf = new byte[4];
+
+            int result = _EVP_EncryptFinal_ex(ctx, outPtr, outLengthBuf);
+
+            outLength = BitConverter.ToInt32(outLengthBuf, 0);
+
+            return result;
+        }
+
+        [DllImport(LibraryName, EntryPoint = "EVP_DecryptInit_ex", SetLastError = true)]
+        private static extern int _EVP_DecryptInit_ex(IntPtr ctx, IntPtr type, IntPtr impl, IntPtr key, IntPtr iv);
+
+        public int EVP_DecryptInit_ex(IntPtr ctx, IntPtr type, IntPtr impl, IntPtr key, IntPtr iv)
+        {
+            return _EVP_DecryptInit_ex(ctx, type, impl, key, iv);
+        }
+
+        [DllImport(LibraryName, EntryPoint = "EVP_DecryptUpdate", SetLastError = true)]
+        private static extern int _EVP_DecryptUpdate(IntPtr ctx, IntPtr outptr, [MarshalAs(UnmanagedType.LPArray)] byte[] outlength, IntPtr inptr, int inlength);
+
+        public int EVP_DecryptUpdate(IntPtr ctx, IntPtr outptr, out int outlength, IntPtr inptr, int inlength)
+        {
+            byte[] outLengthBuf = new byte[4];
+
+            int result = _EVP_DecryptUpdate(ctx, outptr, outLengthBuf, inptr, inlength);
+
+            outlength = BitConverter.ToInt32(outLengthBuf, 0);
+
+            return result;
+        }
+
+        [DllImport(LibraryName, EntryPoint = "EVP_DecryptFinal_ex", SetLastError = true)]
+        private static extern int _EVP_DecryptFinal_ex(IntPtr ctx, IntPtr outPtr, [MarshalAs(UnmanagedType.LPArray)] byte[] outlength);
+
+        public int EVP_DecryptFinal_ex(IntPtr ctx, IntPtr outPtr, out int outLength)
+        {
+            byte[] outLengthBuf = new byte[4];
+
+            int result = _EVP_DecryptFinal_ex(ctx, outPtr, outLengthBuf);
+
+            outLength = BitConverter.ToInt32(outLengthBuf, 0);
+
+            return result;
+        }
+
+        [DllImport(LibraryName, EntryPoint = "RAND_bytes", SetLastError = true)]
+        private static extern int _RAND_bytes(IntPtr buf, int num);
+
+        public int RAND_bytes(IntPtr buf, int num)
+        {
+            return _RAND_bytes(buf, num);
+        }
+
+        [DllImport(LibraryName, EntryPoint = "EVP_CIPHER_CTX_block_size", SetLastError = true)]
+        private static extern int _EVP_CIPHER_CTX_block_size(IntPtr e);
+
+        public int EVP_CIPHER_CTX_block_size(IntPtr e)
+        {
+            return _EVP_CIPHER_CTX_block_size(e);
+        }
+
+        [DllImport(LibraryName, EntryPoint = "EVP_CIPHER_block_size", SetLastError = true)]
+        private static extern int _EVP_CIPHER_block_size(IntPtr e);
+
+        public int EVP_CIPHER_block_size(IntPtr e)
+        {
+            int blockSize = (int)_EVP_CIPHER_block_size(e);
+
+            // BUG: EVP_CIPHER_block_size returns 1
+            if (blockSize == 1)
+            {
+                blockSize = OpenSSLCrypto.EVP_MAX_BLOCK_LENGTH;
+                Debug.WriteLine("BUG: Adjusted block size: " + blockSize);
+            }
```

Comment:
According to their documentation, those maximums are for all supported ciphers.  So this should work regardless.

Reply:
Running a few smoke tests with other ciphers yielded mixed results, though maybe that's more of an issue with the crypto implementation.

To test support for different ciphers I updated the cipher name currently hard-coded in `LinuxOpenSSL11ProtectedMemoryAllocatorLP64` and of the 3 others I tried only chacha20 seemed to work.

```csharp
// cryptProtectMemory = new OpenSSLCryptProtectMemory("chacha20", this);        // Works
// cryptProtectMemory = new OpenSSLCryptProtectMemory("des-ede3-cbc", this);    // Breaks tests
// cryptProtectMemory = new OpenSSLCryptProtectMemory("aes-256-cbc", this);     // Breaks tests
cryptProtectMemory = new OpenSSLCryptProtectMemory("aes-256-gcm", this);        // Works
```

I suppose my concern boils down to multi-cipher support. If this is something we definitely want, then we're going to have to harden up this code a bit, otherwise `OpenSSLCryptProtectMemory`'s cipher probably shouldn't be configurable.

Reply:
From OpenSSL documentation:
The constant EVP_MAX_BLOCK_LENGTH is the maximum block length for all ciphers.

Reply:
Yep, I saw that in the docs as well and I understand why you're using it here as a fallback for `EVP_CIPHER_block_size`, but that's all beside the point; from my previous comment:

> I suppose my concern boils down to multi-cipher support. If this is something we definitely want, then we're going to have to harden up this code a bit, otherwise OpenSSLCryptProtectMemory's cipher probably shouldn't be configurable.

Admittedly, this thread probably wasn't the best place to raise a new concern. Just say the word and we can move this convo elsewhere.

Reply:
> Running a few smoke tests with other ciphers yielded mixed results, though maybe that's more of an issue with the crypto implementation.
> 
> 
> 
> To test support for different ciphers I updated the cipher name currently hard-coded in `LinuxOpenSSL11ProtectedMemoryAllocatorLP64` and of the 3 others I tried only chacha20 seemed to work.
> 
> 
> 
> ```csharp
> 
> // cryptProtectMemory = new OpenSSLCryptProtectMemory("chacha20", this);        // Works
> 
> // cryptProtectMemory = new OpenSSLCryptProtectMemory("des-ede3-cbc", this);    // Breaks tests
> 
> // cryptProtectMemory = new OpenSSLCryptProtectMemory("aes-256-cbc", this);     // Breaks tests
> 
> cryptProtectMemory = new OpenSSLCryptProtectMemory("aes-256-gcm", this);        // Works
> 
> ```
> 
> 
> 
> I suppose my concern boils down to multi-cipher support. If this is something we definitely want, then we're going to have to harden up this code a bit, otherwise `OpenSSLCryptProtectMemory`'s cipher probably shouldn't be configurable.

I want and expected other ciphers to work.  I will investigate this.

Reply:
I figured it out, I needed to do the right rounding for decrypt, but not round on encrypt.

Reply:
All of the ciphers you tested, plus aes-128-cbc work in my testing.